library/src/main/scala/scala/scalajs/runtime/RuntimeLong.scala

```diff
@@ -55,531 +121,558 @@ final class RuntimeLong(
-    else Integer.toBinaryString(l)
+    val lo = this.lo
+    val hi = this.hi
+
+    if (hi != 0) Integer.toBinaryString(hi) + padBinary32(lo)
+    else Integer.toBinaryString(lo)
   }
 
   def toHexString: String = {
-    val zeros = "000000" // 6 zeros
-    @inline def padHex(i: Int, len: Int) = {
+    val zeros = "00000000" // 8 zeros
+    @inline def padBinary8(i: Int) = {
       val s = Integer.toHexString(i)
-      zeros.substring(s.length + (6-len)) + s
+      zeros.substring(s.length) + s
     }
 
-    val mp = m >> 2
-    val lp = l | ((m & 0x3) << BITS)
+    val lo = this.lo
+    val hi = this.hi
 
-    if (h != 0) Integer.toHexString(h) + padHex(mp, 5) + padHex(lp, 6)
-    else if (mp != 0) Integer.toHexString(mp) + padHex(lp, 6)
-    else Integer.toHexString(lp)
+    if (hi != 0) Integer.toHexString(hi) + padBinary8(lo)
+    else Integer.toHexString(lo)
   }
 
   def toOctalString: String = {
-    val zeros = "0000000" // 7 zeros
-    @inline def padOctal7(i: Int) = {
+    val zeros = "0000000000" // 10 zeros
+    @inline def padOctal10(i: Int) = {
       val s = Integer.toOctalString(i)
       zeros.substring(s.length) + s
     }
 
-    val lp = l & (MASK >> 1)
-    val mp = ((m & (MASK >> 2)) << 1) | (l >> (BITS - 1))
-    val hp = (h << 2) | (m >> (BITS - 2))
+    val lo = this.lo
+    val hi = this.hi
+
+    val lp = lo & 0x3fffffff
+    val mp = ((lo >>> 30) + (hi << 2)) & 0x3fffffff
+    val hp = hi >>> 28
 
-    if (hp != 0) Integer.toOctalString(hp) + padOctal7(mp) + padOctal7(lp)
-    else if (mp != 0) Integer.toOctalString(mp) + padOctal7(lp)
+    if (hp != 0) Integer.toOctalString(hp) + padOctal10(mp) + padOctal10(lp)
+    else if (mp != 0) Integer.toOctalString(mp) + padOctal10(lp)
     else Integer.toOctalString(lp)
   }
 
-  // Any API //
-
-  override def toString: String = {
-    if (isZero) "0"
-    // Check for MinValue, because its not negatable
-    else if (isMinValue) "-9223372036854775808"
-    else if (isNegative) "-" + (-x).toString
-    else {
-      val tenPow9 = TenPow9 // local copy to access CachedConstants only once
-
-      @tailrec
-      @inline
-      def toString0(v: RuntimeLong, acc: String): String =
-        if (v.isZero) acc
-        else {
-          val quotRem = v.divMod(tenPow9)
-          val quot = quotRem(0)
-          val rem = quotRem(1)
+  def bitCount: Int =
+    Integer.bitCount(lo) + Integer.bitCount(hi)
 
-          val digits = rem.toInt.toString
-          val zeroPrefix =
-            if (quot.isZero) ""
-            else "000000000".substring(digits.length) // (9 - digits.length) zeros
+  def signum: RuntimeLong = {
+    val hi = this.hi
+    if (hi < 0) MinusOne
+    else if (isZero(lo, hi)) Zero
+    else One
+  }
 
-          toString0(quot, zeroPrefix + digits + acc)
-        }
+  def numberOfLeadingZeros: Int = {
+    val hi = this.hi
+    if (hi != 0) Integer.numberOfLeadingZeros(hi)
+    else Integer.numberOfLeadingZeros(lo) + 32
+  }
```

Comment:
`inlineNumberOfLeadingZeros` forces to read `lo` from the heap even if it is not needed.